ra2ce/network/avg_speed/avg_speed_calculator.py

```diff
@@ -0,0 +1,215 @@
+"""
+                    GNU GENERAL PUBLIC LICENSE
+                      Version 3, 29 June 2007
+    Risk Assessment and Adaptation for Critical Infrastructure (RA2CE).
+    Copyright (C) 2023 Stichting Deltares
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+import logging
+from pathlib import Path
+from re import split
+from statistics import mean
+
+import networkx as nx
+
+import ra2ce.network.networks_utils as nut
+from ra2ce.network.avg_speed.avg_speed import AvgSpeed
+from ra2ce.network.avg_speed.avg_speed_reader import AvgSpeedReader
+from ra2ce.network.avg_speed.avg_speed_writer import AvgSpeedWriter
+from ra2ce.network.network_config_data.enums.road_type_enum import RoadTypeEnum
+
+
+class AvgSpeedCalculator:
+    """
+    Class to calculate the average speed of a road network based on the maximum speed of the roads.
+    """
+
+    graph: nx.Graph
+    output_graph_dir: Path | None
+    avg_speed: AvgSpeed
+
+    def __init__(self, graph: nx.Graph, output_graph_dir: Path | None) -> None:
+        self.graph = graph
+        self.output_graph_dir = output_graph_dir
+        self.avg_speed = None
+
+    @staticmethod
+    def parse_speed(speed_input: str | list[str]) -> float:
+        """
+        Parse the average speed from the input string(s).
+
+        Args:
+            speed_input (str | list[str]): (List of) string(s) with the speed(s).
+                Can have different formats, e.g. "50 mph", "50", "50;60", "50-60", "50|60".
+
+        Returns:
+            float: Average speed of the input string(s).
+                0.0 is returned if the input is empty or can't be parsed correctly.
+        """
+        if not speed_input:
+            return 0.0
+        if isinstance(speed_input, list):
+            return mean(map(AvgSpeedCalculator.parse_speed, speed_input))
+        if " mph" in speed_input:
+            return float(speed_input.split(" mph")[0]) * 1.609344
+        if any(c.isalpha() for c in speed_input):
+            return 0.0
+        # Split on the different separators (";" or "-" or "|")
+        # and take the mean of the results (if any).
+        return mean(float(x) for x in split(r";|-|\|", speed_input) if x.isnumeric())
```

Comment:
This is great, but one remark, recursion can become a sinkhole (potential infinite loops) and give you problems. You are  (usually) better off by 'mapping' all the possibilities into something you can get the grip of, for in stance making sure all your values are a list of floats where the `mph` has already been checked.

The last if ( `if any(c.isalpha()`) is a bit confusing, please specify why its required.

Reply:
In general I agree to the recursion statement. In this case, however, it goes only 1 level deeper.
2b sure I flattened the list first:
```
        if isinstance(speed_input, list):
            _flattened_list = list(chain.from_iterable(speed_input))
            return mean(map(AvgSpeedCalculator.parse_speed, _flattened_list))
```

`if any(c.isalpha())` was copied from the original code and is meant to check if the string can be parsed as float later on. That is not the case if it contains letters, hence `isalpha()`.

Reply:
Correction: I couldn't find a (non-recursive) way to flatten the list before calling the recursion.

Reply:
Agreed to leave it like this